src/test/java/gov/cdc/sdp/cbr/queue/DatabaseQueueParallelTest.java

```diff
@@ -0,0 +1,95 @@
+package gov.cdc.sdp.cbr.queue;
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.sql.DataSource;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.EndpointInject;
+import org.apache.camel.Exchange;
+import org.apache.camel.Message;
+import org.apache.camel.Produce;
+import org.apache.camel.ProducerTemplate;
+import org.apache.camel.component.mock.MockEndpoint;
+import org.apache.camel.impl.DefaultExchange;
+import org.apache.camel.impl.DefaultMessage;
+import org.apache.camel.test.spring.CamelSpringJUnit4ClassRunner;
+import org.apache.camel.test.spring.CamelTestContextBootstrapper;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.annotation.Repeat;
+import org.springframework.test.context.BootstrapWith;
+import org.springframework.test.context.ContextConfiguration;
+
+@RunWith(CamelSpringJUnit4ClassRunner.class)
+@BootstrapWith(CamelTestContextBootstrapper.class)
+@ContextConfiguration(locations = { "classpath:DatabaseQueueParallelTest-context.xml" })
+@PropertySource("classpath:application.properties")
+public class DatabaseQueueParallelTest {
+
+	@Autowired
+	protected CamelContext camelContext;
+
+	@EndpointInject(uri = "mock:mock_endpoint")
+	protected MockEndpoint mock_endpoint;
+
+	@EndpointInject(uri = "mock:mock_endpoint2")
+	protected MockEndpoint mock_endpoint2;
+
+	@EndpointInject(uri = "mock:mock_endpoint3")
+	protected MockEndpoint mock_endpoint3;
+
+	@Before
+	public void setUp() {
+		DataSource ds = (DataSource) camelContext.getRegistry().lookupByName("sdpqDataSource");
+		JdbcTemplate jdbcTemplate = new JdbcTemplate(ds);
+
+		String minimum_required_headers = "(cbr_id, source, source_id, payload, cbr_recevied_time)";
+		String col_val = minimum_required_headers
+				+ " values ('cbr_1337', 'mockland', 'mockland_1', 'the payload', '"
+				+ new Date(System.currentTimeMillis()) + "')";
+		String tableName = "message_queue";
+
+		String create_dummy_data = "INSERT into " + tableName + col_val;
+		
+		for (int i=0; i<1000; i++) {
+			jdbcTemplate.update(create_dummy_data);
+		}
+
+	}
+	
+	@Produce(uri = "direct:start")
+	protected ProducerTemplate template;
+
+	@Test
+	@DirtiesContext
+	public void testQueueConsumer() throws Exception {
```

Comment:
How does this test that concurrency isn't an issue?  I only see one endpoint being tested.

Reply:
The test is inserting records into the database.  There are two routes both running in the context, both of which use a database queue component to consume the records.  Both routes are sending to the same endpoint, so regardless of which route processes them, they will arrive at that endpoint.  If both routes process the same message, it will arrive twice, and more than 1000 messages will reach the endpoint.

When this test is run without the changes to DatabaseQueueConsumer, the endpoint receives something more than 1000 messages -- between 1015 and 1100, usually.